extension-iap/src/java/com/defold/iap/IapGooglePlay.java

```diff
@@ -356,30 +414,43 @@ public void onSkuDetailsResponse(BillingResult billingResult, List<SkuDetails> s
      * Get details for a list of products. The products can be a mix of
      * in-app products and subscriptions.
      */
-    private void querySkuDetailsAsync(final List<String> skuList, final SkuDetailsResponseListener listener) {
-        SkuDetailsResponseListener detailsListener = new SkuDetailsResponseListener() {
-            private List<SkuDetails> allSkuDetails = new ArrayList<SkuDetails>();
+    private void queryProductDetailsAsync(final List<String> productList, final ProductDetailsResponseListener listener) {
+        ProductDetailsResponseListener detailsListener = new ProductDetailsResponseListener() {
+            private List<ProductDetails> allProductDetails = new ArrayList<ProductDetails>();
             private int queries = 2;
 
             @Override
-            public void onSkuDetailsResponse(BillingResult billingResult, List<SkuDetails> skuDetails) {
-                if (skuDetails != null) {
-                    // cache skus (cache will be used to speed up buying)
-                    for (SkuDetails sd : skuDetails) {
-                        IapGooglePlay.this.products.put(sd.getSku(), sd);
+            public void onProductDetailsResponse(BillingResult billingResult, List<ProductDetails> productDetails) {
+                if (productDetails != null) {
+                    // cache products (cache will be used to speed up buying)
+                    for (ProductDetails pd : productDetails) {
+                        IapGooglePlay.this.products.put(pd.getProductId(), pd);
                     }
-                    // add to list of all sku details
-                    allSkuDetails.addAll(skuDetails);
+                    // add to list of all product details
+                    allProductDetails.addAll(productDetails);
                 }
                 // we're finished when we have queried for both in-app and subs
                 queries--;
                 if (queries == 0) {
-                    listener.onSkuDetailsResponse(billingResult, allSkuDetails);
+                    listener.onProductDetailsResponse(billingResult, allProductDetails);
                 }
             }
         };
-        billingClient.querySkuDetailsAsync(SkuDetailsParams.newBuilder().setSkusList(skuList).setType(SkuType.INAPP).build(), detailsListener);
-        billingClient.querySkuDetailsAsync(SkuDetailsParams.newBuilder().setSkusList(skuList).setType(SkuType.SUBS).build(), detailsListener);
+
+        // we don't know if a product is a subscription or inapp product type
+        // instread we create two product lists from the same set of products and use INAPP for one and SUBS for the other
```

Comment:
hmm, is it valid way to do it? Shouldn't we ask user to specify it?

Reply:
This is the way we did it in previous versions of the extension as well.